File: src/com/Torvald/Terrarum/Game.java
package com.Torvald.Terrarum;

import com.Torvald.Rand.HQRNG;
import com.Torvald.Terrarum.Actors.*;
import com.Torvald.Terrarum.ConsoleCommand.Authenticator;
import com.Torvald.Terrarum.ConsoleCommand.CommandDict;
import com.Torvald.Terrarum.GameControl.GameController;
import com.Torvald.Terrarum.GameControl.KeyMap;
import com.Torvald.Terrarum.GameMap.GameMap;
import com.Torvald.Terrarum.GameMap.WorldTime;
import com.Torvald.Terrarum.MapDrawer.LightmapRenderer;
import com.Torvald.Terrarum.MapDrawer.MapCamera;
import com.Torvald.Terrarum.MapDrawer.MapDrawer;
import com.Torvald.Terrarum.MapGenerator.MapGenerator;
import com.Torvald.Terrarum.MapGenerator.RoguelikeRandomiser;
import com.Torvald.Terrarum.TileProperties.TilePropCodex;
import com.Torvald.Terrarum.TileStat.TileStat;
import com.Torvald.Terrarum.UserInterface.*;
import com.jme3.math.FastMath;
import org.lwjgl.opengl.ARBShaderObjects;
import org.lwjgl.opengl.GL11;
import org.newdawn.slick.*;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.fills.GradientFill;
import org.newdawn.slick.geom.Rectangle;
import org.newdawn.slick.state.BasicGameState;
import org.newdawn.slick.state.StateBasedGame;
import shader.Shader;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.lang.management.ManagementFactory;
import java.util.HashSet;

/**
 * Created by minjaesong on 15-12-30.
 */
public class Game extends BasicGameState {

    public static long memInUse;
    public static long totalVMMem;
    int game_mode = 0;

    public GameMap map;

    public HashSet<Actor> actorContainer = new HashSet<>();
    public HashSet<UIHandler> uiContainer = new HashSet<>();

    public UIHandler consoleHandler;
    public UIHandler debugWindow;
    public UIHandler notifinator;

    @NotNull Player player;

    private Image GRADIENT_IMAGE;
    private Rectangle skyBox;

    public float screenZoom = 1.0f;
    public final float ZOOM_MAX = 2.0f;
    public final float ZOOM_MIN = 0.25f;

    private Shader shader12BitCol;
    private Shader shaderBlurH;
    private Shader shaderBlurV;

    public static Authenticator auth = new Authenticator();

    public Game() throws SlickException {  }


    private boolean useShader;
    private int shaderProgram = 0;


    private final int ENV_COLTEMP_SUNRISE = 2500;
    private final int ENV_SUNLIGHT_DELTA = MapDrawer.getENV_COLTEMP_NOON() - ENV_COLTEMP_SUNRISE;


    @Override
    public void init(GameContainer gameContainer, StateBasedGame stateBasedGame) throws
            SlickException {
        new GameController();
        KeyMap.build();
        GameController.setKeyMap(new KeyMap());


        shader12BitCol = Shader.makeShader("./res/4096.vrt", "./res/4096.frg");
        shaderBlurH = Shader.makeShader("./res/blurH.vrt", "./res/blur.frg");
        shaderBlurV = Shader.makeShader("./res/blurV.vrt", "./res/blur.frg");


        GRADIENT_IMAGE = new Image("res/graphics/sky_colour.png");
        skyBox = new Rectangle(0, 0, Terrarum.WIDTH, Terrarum.HEIGHT);

        new WorldTime();
        new TilePropCodex();
        // new ItemPropCodex() -- This is kotlin object and already initialised.

        map = new GameMap(8192, 2048);
        map.setGravitation(9.8f);

        MapGenerator.attachMap(map);
        MapGenerator.setSeed(0x51621D);
        //MapGenerator.setSeed(new HQRNG().nextLong());
        MapGenerator.generateMap();

        RoguelikeRandomiser.setSeed(0x540198);
        //RoguelikeRandomiser.setSeed(new HQRNG().nextLong());


        new CommandDict();

        // add new player and put it to actorContainer
        //player = new Player();
        player = PFSigrid.build();
        //player.setNoClip(true);
        actorContainer.add(player);

        consoleHandler = new UIHandler(new ConsoleWindow());
        consoleHandler.setPosition(0, 0);

        debugWindow = new UIHandler(new BasicDebugInfoWindow());
        debugWindow.setPosition(0, 0);

        notifinator = new UIHandler(new Notification());
        notifinator.setPosition(
                (Terrarum.WIDTH - notifinator.getUI().getWidth())
                        / 2
                , Terrarum.HEIGHT - notifinator.getUI().getHeight()
        );
        notifinator.setVisibility(true);
    }

    public Player getPlayer() {
        return player;
    }

    @Override
    public void update(GameContainer gc, StateBasedGame sbg, int delta_t) {
        setAppTitle();

        // GL at after_sunrise-noon_before_sunset
        //map.setGlobalLight();

        GameController.processInput(gc.getInput());

        TileStat.update();

        MapDrawer.update(gc, delta_t);
        MapCamera.update(gc, delta_t);

        actorContainer.forEach(actor -> actor.update(gc, delta_t));
        actorContainer.forEach(
                actor -> {
                    if (actor instanceof Visible) {
                        ((Visible) actor).updateBodySprite(gc, delta_t);
                    }
                    if (actor instanceof Glowing) {
                        ((Glowing) actor).updateGlowSprite(gc, delta_t);
                    }
                }
        );

        uiContainer.forEach(ui -> ui.update(gc, delta_t));

        notifinator.update(gc, delta_t);

        Terrarum.appgc.setVSync(Terrarum.appgc.getFPS() >= Terrarum.VSYNC_TRIGGER_THRESHOLD);
    }

    private void setAppTitle() {
        Runtime runtime = Runtime.getRuntime();
        memInUse = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage().getUsed() >> 20;
        totalVMMem = runtime.maxMemory() >> 20;

        Terrarum.appgc.setTitle(
                "Simple Slick Game — FPS: "
                        + Terrarum.appgc.getFPS() + " ("
                        + String.valueOf(Terrarum.TARGET_INTERNAL_FPS)
                        + ") — "
                        + String.valueOf(memInUse) + "M / "
                        + String.valueOf(totalVMMem) + "M"
        );
    }

    @Override
    public void render(GameContainer gc, StateBasedGame sbg, Graphics g) {

        drawSkybox(g);

        // compensate for zoom. UIs have to be treated specially! (see UIHandler)
        g.translate(
                  -MapCamera.getCameraX() * screenZoom
                , -MapCamera.getCameraY() * screenZoom
        );

        MapCamera.renderBehind(gc, g);

        actorContainer.forEach(
                actor -> { if (actor instanceof Visible) ((Visible) actor).drawBody(gc, g); }
        );
        actorContainer.forEach(
                actor -> { if (actor instanceof Glowing) ((Glowing) actor).drawGlow(gc, g); }
        );

        LightmapRenderer.renderLightMap();

        MapCamera.renderFront(gc, g);
        MapDrawer.render(gc, g);

        setBlendModeMul();
        MapDrawer.drawEnvOverlay(g);
        LightmapRenderer.draw(g);
        setBlendModeNormal();

        uiContainer.forEach(ui -> ui.render(gc, g));
        debugWindow.render(gc, g);
        consoleHandler.render(gc, g);
        notifinator.render(gc, g);
    }

    public boolean addActor(Actor e) {
        return actorContainer.add(e);
    }

    public boolean removeActor(Actor e) {
        return actorContainer.remove(e);
    }

    private Color[] getGradientColour(int timeSec) {
        Color[] colourTable = new Color[2];

        int gradMapWidth = GRADIENT_IMAGE.getWidth();
        int phase = Math.round((timeSec / WorldTime.DAY_LENGTH) * gradMapWidth);

        //update in every INTERNAL_FRAME frames
        colourTable[0] = GRADIENT_IMAGE.getColor(phase, 0);
        colourTable[1] = GRADIENT_IMAGE.getColor(phase, 1);

        return colourTable;
    }

    public void keyPressed(int key, char c) {
        GameController.keyPressed(key, c);
    }

    public void keyReleased(int key, char c) {
        GameController.keyReleased(key, c);
    }

    public void mouseMoved(int oldx, int oldy, int newx, int newy) {
        GameController.mouseMoved(oldx, oldy, newx, newy);
    }

    public void mouseDragged(int oldx, int oldy, int newx, int newy) {
        GameController.mouseDragged(oldx, oldy, newx, newy);
    }

    public void mousePressed(int button, int x, int y) {
        GameController.mousePressed(button, x, y);
    }

    public void mouseReleased(int button, int x, int y) {
        GameController.mouseReleased(button, x, y);
    }

    public void mouseWheelMoved(int change) {
        GameController.mouseWheelMoved(change);
    }

    public void controllerButtonPressed(int controller, int button) {
        GameController.controllerButtonPressed(controller, button);
    }

    public void controllerButtonReleased(int controller, int button) {
        GameController.controllerButtonReleased(controller, button);
    }

    @Override
    public int getID() {
        return Terrarum.SCENE_ID_GAME;
    }

    private void drawSkybox(Graphics g) {
        Color[] colourTable = getGradientColour(Terrarum.game.map.getWorldTime().elapsedSeconds());
        GradientFill skyColourFill = new GradientFill(0, 0, colourTable[0], 0, Terrarum.HEIGHT, colourTable[1]);
        g.fill(skyBox, skyColourFill);
    }

    private void setBlendModeMul() {
        GL11.glEnable(GL11.GL_BLEND);
        GL11.glBlendFunc(GL11.GL_DST_COLOR, GL11.GL_ONE_MINUS_SRC_ALPHA);
    }

    private void setBlendModeNormal() {
        GL11.glDisable(GL11.GL_BLEND);
        Terrarum.appgc.getGraphics().setDrawMode(Graphics.MODE_NORMAL);
    }

    public void sendNotification(String[] msg) {
        ((Notification) notifinator.getUI()).sendNotification(msg);
    }

    private int createShader(String filename, int shaderType) throws Exception {
        int shader = 0;
        try {
            shader = ARBShaderObjects.glCreateShaderObjectARB(shaderType);

            if(shader == 0)
                return 0;

            ARBShaderObjects.glShaderSourceARB(shader, readFileAsString(filename));
            ARBShaderObjects.glCompileShaderARB(shader);

            if (ARBShaderObjects.glGetObjectParameteriARB(shader, ARBShaderObjects.GL_OBJECT_COMPILE_STATUS_ARB) == GL11.GL_FALSE)
                throw new RuntimeException("Error creating shader: " + getLogInfo(shader));

            return shader;
        }
        catch(Exception exc) {
            ARBShaderObjects.glDeleteObjectARB(shader);
            throw exc;
        }
    }

    private static String getLogInfo(int obj) {
        return ARBShaderObjects.glGetInfoLogARB(obj, ARBShaderObjects.glGetObjectParameteriARB(obj, ARBShaderObjects.GL_OBJECT_INFO_LOG_LENGTH_ARB));
    }

    private String readFileAsString(String filename) throws Exception {
        StringBuilder source = new StringBuilder();

        FileInputStream in = new FileInputStream(filename);

        Exception exception = null;

        BufferedReader reader;
        try{
            reader = new BufferedReader(new InputStreamReader(in,"UTF-8"));

            Exception innerExc= null;
            try {
                String line;
                while((line = reader.readLine()) != null)
                    source.append(line).append('\n');
            }
            catch(Exception exc) {
                exception = exc;
            }
            finally {
                try {
                    reader.close();
                }
                catch(Exception exc) {
                    if(innerExc == null)
                        innerExc = exc;
                    else
                        exc.printStackTrace();
                }
            }

            if(innerExc != null)
                throw innerExc;
        }
        catch(Exception exc) {
            exception = exc;
        }
        finally {
            try {
                in.close();
            }
            catch(Exception exc) {
                if(exception == null)
                    exception = exc;
                else
                    exc.printStackTrace();
            }

            if(exception != null)
                throw exception;
        }

        return source.toString();
    }

    public long getMemInUse() {
        return memInUse;
    }

    public long getTotalVMMem() {
        return totalVMMem;
    }

    private int getSunlightColtemp() {
        int half_today = WorldTime.DAY_LENGTH / 2;
        int timeToday = WorldTime.elapsedSeconds();
        float sunAlt = (timeToday < half_today) ?
                        timeToday / half_today * FastMath.PI
                       : 0f;
        return Math.round(ENV_COLTEMP_SUNRISE + (ENV_SUNLIGHT_DELTA * FastMath.sin(sunAlt)));
    }
}

File: src/com/Torvald/Terrarum/Game.kt
package com.torvald.terrarum

import com.torvald.terrarum.gameactors.*
import com.torvald.terrarum.console.Authenticator
import com.torvald.terrarum.gamecontroller.GameController
import com.torvald.terrarum.gamecontroller.Key
import com.torvald.terrarum.gamecontroller.KeyMap
import com.torvald.terrarum.gamecontroller.KeyToggler
import com.torvald.terrarum.gamemap.GameMap
import com.torvald.terrarum.gamemap.WorldTime
import com.torvald.terrarum.mapdrawer.LightmapRenderer
import com.torvald.terrarum.mapdrawer.MapCamera
import com.torvald.terrarum.mapdrawer.MapDrawer
import com.torvald.terrarum.mapgenerator.MapGenerator
import com.torvald.terrarum.mapgenerator.RoguelikeRandomiser
import com.torvald.terrarum.tileproperties.TilePropCodex
import com.torvald.terrarum.tilestats.TileStats
import com.torvald.terrarum.ui.BasicDebugInfoWindow
import com.torvald.terrarum.ui.ConsoleWindow
import com.torvald.terrarum.ui.Notification
import com.torvald.terrarum.ui.UIHandler
import org.lwjgl.opengl.GL11
import org.newdawn.slick.*
import org.newdawn.slick.fills.GradientFill
import org.newdawn.slick.geom.Rectangle
import org.newdawn.slick.state.BasicGameState
import org.newdawn.slick.state.StateBasedGame
import shader.Shader
import java.lang.management.ManagementFactory
import java.util.*

/**
 * Created by minjaesong on 16-03-19.
 */
class Game @Throws(SlickException::class)
constructor() : BasicGameState() {
    internal var game_mode = 0

    lateinit var map: GameMap

    var actorContainer = HashSet<Actor>()
    var uiContainer = HashSet<UIHandler>()

    lateinit var consoleHandler: UIHandler
    lateinit var debugWindow: UIHandler
    lateinit var notifinator: UIHandler

    lateinit internal var player: Player

    private var GRADIENT_IMAGE: Image? = null
    private var skyBox: Rectangle? = null

    var screenZoom = 1.0f
    val ZOOM_MAX = 2.0f
    val ZOOM_MIN = 0.25f

    private var shader12BitCol: Shader? = null
    private var shaderBlurH: Shader? = null
    private var shaderBlurV: Shader? = null


    private val useShader: Boolean = false
    private val shaderProgram = 0


    private val ENV_COLTEMP_SUNRISE = 2500
    private val ENV_SUNLIGHT_DELTA = MapDrawer.ENV_COLTEMP_NOON - ENV_COLTEMP_SUNRISE


    val memInUse: Long
        get() = ManagementFactory.getMemoryMXBean().heapMemoryUsage.used shr 20
    val totalVMMem: Long
        get() = Runtime.getRuntime().maxMemory() shr 20

    val auth = Authenticator()

    private var update_delta: Int = 0

    val KEY_LIGHTMAP_RENDER = Key.F7
    val KEY_LIGHTMAP_SMOOTH = Key.F8

    @Throws(SlickException::class)
    override fun init(gameContainer: GameContainer, stateBasedGame: StateBasedGame) {
        KeyMap.build()

        shader12BitCol = Shader.makeShader("./res/4096.vrt", "./res/4096.frg")
        shaderBlurH = Shader.makeShader("./res/blurH.vrt", "./res/blur.frg")
        shaderBlurV = Shader.makeShader("./res/blurV.vrt", "./res/blur.frg")


        GRADIENT_IMAGE = Image("res/graphics/sky_colour.png")
        skyBox = Rectangle(0f, 0f, Terrarum.WIDTH.toFloat(), Terrarum.HEIGHT.toFloat())

        TilePropCodex()
        // new ItemPropCodex() -- This is kotlin object and already initialised.

        map = GameMap(8192, 2048)
        map.gravitation = 9.8f

        MapGenerator.attachMap(map)
        MapGenerator.setSeed(0x51621D2)
        //mapgenerator.setSeed(new HQRNG().nextLong());
        MapGenerator.generateMap()

        RoguelikeRandomiser.setSeed(0x540198)
        //RoguelikeRandomiser.setSeed(new HQRNG().nextLong());


        // add new player and put it to actorContainer
        player = PFSigrid.create()
        //player = PFCynthia.create()
        //player.setNoClip(true);
        actorContainer.add(player)

        consoleHandler = UIHandler(ConsoleWindow())
        consoleHandler.setPosition(0, 0)

        debugWindow = UIHandler(BasicDebugInfoWindow())
        debugWindow.setPosition(0, 0)

        notifinator = UIHandler(Notification())
        notifinator.setPosition(
                (Terrarum.WIDTH - notifinator.UI.width) / 2, Terrarum.HEIGHT - notifinator.UI.height)
        notifinator.setVisibility(true)

        if (Terrarum.gameConfig.getAsBoolean("smoothlighting") == true)
            KeyToggler.forceSet(KEY_LIGHTMAP_SMOOTH, true)
        else
            KeyToggler.forceSet(KEY_LIGHTMAP_SMOOTH, false)
    }

    override fun update(gc: GameContainer, sbg: StateBasedGame, delta: Int) {
        update_delta = delta
        setAppTitle()

        // GL at after_sunrise-noon_before_sunset
        map.updateWorldTime(delta)
        map.globalLight = globalLightByTime

        GameController.processInput(gc.input)

        TileStats.update()

        MapDrawer.update(gc, delta)
        MapCamera.update(gc, delta)

        actorContainer.forEach { actor -> actor.update(gc, delta) }
        actorContainer.forEach { actor ->
            if (actor is Visible) {
                actor.updateBodySprite(gc, delta)
            }
            if (actor is Glowing) {
                actor.updateGlowSprite(gc, delta)
            }
        }

        uiContainer.forEach { ui -> ui.update(gc, delta) }
        consoleHandler.update(gc, delta)
        debugWindow.update(gc, delta)


        notifinator.update(gc, delta)

        Terrarum.appgc.setVSync(Terrarum.appgc.fps >= Terrarum.VSYNC_TRIGGER_THRESHOLD)
    }

    private fun setAppTitle() {
        Terrarum.appgc.setTitle(
                "Simple Slick Game — FPS: "
                + Terrarum.appgc.fps + " ("
                + Terrarum.TARGET_INTERNAL_FPS.toString()
                + ") — "
                + memInUse.toString() + "M / "
                + totalVMMem.toString() + "M")
    }

    override fun render(gc: GameContainer, sbg: StateBasedGame, g: Graphics) {
        Terrarum.gameConfig["smoothlighting"] = KeyToggler.isOn(KEY_LIGHTMAP_SMOOTH)

        if (!g.isAntiAlias) g.isAntiAlias = true

        drawSkybox(g)

        // compensate for zoom. UIs have to be treated specially! (see UIHandler)
        g.translate(
                -MapCamera.cameraX * screenZoom, -MapCamera.cameraY * screenZoom)

        MapCamera.renderBehind(gc, g)

        actorContainer.forEach { actor ->
            if (actor is Visible) actor.drawBody(gc, g)
            if (actor is Glowing) actor.drawGlow(gc, g)
        }

        player.drawBody(gc, g)
        player.drawGlow(gc, g)

        LightmapRenderer.renderLightMap()

        MapCamera.renderFront(gc, g)
        MapDrawer.render(gc, g)


        setBlendModeMul()

        MapDrawer.drawEnvOverlay(g)

        if (!KeyToggler.isOn(KEY_LIGHTMAP_RENDER)) setBlendModeMul()
        else setBlendModeNormal()

        LightmapRenderer.draw(g)

        setBlendModeNormal()

        uiContainer.forEach { ui -> ui.render(gc, g) }
        debugWindow.render(gc, g)
        consoleHandler.render(gc, g)
        notifinator.render(gc, g)
    }

    fun addActor(e: Actor): Boolean {
        return actorContainer.add(e)
    }

    fun removeActor(e: Actor): Boolean {
        return actorContainer.remove(e)
    }

    private fun getGradientColour(row: Int): Color {
        val gradMapWidth = GRADIENT_IMAGE!!.width
        val phase = Math.round(
                map.worldTime.elapsedSeconds().toFloat() / WorldTime.DAY_LENGTH.toFloat() * gradMapWidth
        )

        //update in every INTERNAL_FRAME frames
        return GRADIENT_IMAGE!!.getColor(phase, row)
    }

    override fun keyPressed(key: Int, c: Char) {
        GameController.keyPressed(key, c)
    }

    override fun keyReleased(key: Int, c: Char) {
        GameController.keyReleased(key, c)
    }

    override fun mouseMoved(oldx: Int, oldy: Int, newx: Int, newy: Int) {
        GameController.mouseMoved(oldx, oldy, newx, newy)
    }

    override fun mouseDragged(oldx: Int, oldy: Int, newx: Int, newy: Int) {
        GameController.mouseDragged(oldx, oldy, newx, newy)
    }

    override fun mousePressed(button: Int, x: Int, y: Int) {
        GameController.mousePressed(button, x, y)
    }

    override fun mouseReleased(button: Int, x: Int, y: Int) {
        GameController.mouseReleased(button, x, y)
    }

    override fun mouseWheelMoved(change: Int) {
        GameController.mouseWheelMoved(change)
    }

    override fun controllerButtonPressed(controller: Int, button: Int) {
        GameController.controllerButtonPressed(controller, button)
    }

    override fun controllerButtonReleased(controller: Int, button: Int) {
        GameController.controllerButtonReleased(controller, button)
    }

    override fun getID(): Int {
        return Terrarum.SCENE_ID_GAME
    }

    private fun drawSkybox(g: Graphics) {
        val skyColourFill = GradientFill(
                0f, 0f, getGradientColour(0),
                0f, Terrarum.HEIGHT.toFloat(), getGradientColour(1)
        )
        g.fill(skyBox, skyColourFill)
    }

    private fun setBlendModeMul() {
        GL11.glEnable(GL11.GL_BLEND)
        GL11.glBlendFunc(GL11.GL_DST_COLOR, GL11.GL_ONE_MINUS_SRC_ALPHA)
    }

    private fun setBlendModeNormal() {
        GL11.glDisable(GL11.GL_BLEND)
        Terrarum.appgc.graphics.setDrawMode(Graphics.MODE_NORMAL)
    }

    fun sendNotification(msg: Array<String>) {
        (notifinator.UI as Notification).sendNotification(Terrarum.appgc, update_delta, msg)
        notifinator.setAsOpening()
    }

    private val globalLightByTime: Int
        get() = getGradientColour(2).getRGB24()

    /**
     * extension function for org.newdawn.slick.Color
     */
    fun Color.getRGB24(): Int = (this.redByte shl 16) or (this.greenByte shl 8) or (this.blueByte)
}
